### PR TITLE
Upgrade dotnet to v6 + list_of_files for dotnet-format

### PR DESCRIPTION
**Pull request by nvuillam**
Fixes https://github.com/oxsecurity/megalinter/issues/1642

### CODE REVIEW

**Comment by nvuillam** [comment]
## [MegaLinter](https://oxsecurity.github.io/megalinter/latest) status: :warning: [**WARNING**](https://github.com/oxsecurity/megalinter/actions/runs/2728373587)

|     Descriptor     |                                                         Linter                                                         |Files|Fixed|                                  Errors                                   |Elapsed time|
|--------------------|------------------------------------------------------------------------------------------------------------------------|-----|----:|---------------------------------------------------------------------------|------------|
|✅ BASH             |[bash-exec](https://oxsecurity.github.io/megalinter/latest/descriptors/bash_bash_exec)                                  |    7|     |                                                                          0|0.05s       |
|✅ BASH             |[shellcheck](https://oxsecurity.github.io/megalinter/latest/descriptors/bash_shellcheck)                                |    7|     |                                                                          0|0.28s       |
|✅ BASH             |[shfmt](https://oxsecurity.github.io/megalinter/latest/descriptors/bash_shfmt)                                          |    7|    0|                                                                          0|0.38s       |
|✅ COPYPASTE        |[jscpd](https://oxsecurity.github.io/megalinter/latest/descriptors/copypaste_jscpd)                                     |yes  |     |no                                                                         |2.44s       |
|✅ DOCKERFILE       |[hadolint](https://oxsecurity.github.io/megalinter/latest/descriptors/dockerfile_hadolint)                              |  102|     |                                                                          0|19.25s      |
|✅ JSON             |[eslint-plugin-jsonc](https://oxsecurity.github.io/megalinter/latest/descriptors/json_eslint_plugin_jsonc)              |   20|    0|                                                                          0|2.69s       |
|✅ JSON             |[jsonlint](https://oxsecurity.github.io/megalinter/latest/descriptors/json_jsonlint)                                    |   18|     |                                                                          0|5.58s       |
|✅ JSON             |[v8r](https://oxsecurity.github.io/megalinter/latest/descriptors/json_v8r)                                              |   20|     |                                                                          0|11.97s      |
|:warning: MARKDOWN  |[markdownlint](https://oxsecurity.github.io/megalinter/latest/descriptors/markdown_markdownlint)                        |  279|    0|[**4**](https://github.com/oxsecurity/megalinter/actions/runs/2728373587)  |5.36s       |
|✅ MARKDOWN         |[markdown-link-check](https://oxsecurity.github.io/megalinter/latest/descriptors/markdown_markdown_link_check)          |  279|     |                                                                          0|5.34s       |
|✅ MARKDOWN         |[markdown-table-formatter](https://oxsecurity.github.io/megalinter/latest/descriptors/markdown_markdown_table_formatter)|  279|    0|                                                                          0|6.85s       |
|✅ OPENAPI          |[spectral](https://oxsecurity.github.io/megalinter/latest/descriptors/openapi_spectral)                                 |    1|     |                                                                          0|1.04s       |
|:warning: PYTHON    |[bandit](https://oxsecurity.github.io/megalinter/latest/descriptors/python_bandit)                                      |  163|     |[**50**](https://github.com/oxsecurity/megalinter/actions/runs/2728373587) |2.44s       |
|✅ PYTHON           |[black](https://oxsecurity.github.io/megalinter/latest/descriptors/python_black)                                        |  163|    0|                                                                          0|4.18s       |
|✅ PYTHON           |[flake8](https://oxsecurity.github.io/megalinter/latest/descriptors/python_flake8)                                      |  163|     |                                                                          0|2.86s       |
|✅ PYTHON           |[isort](https://oxsecurity.github.io/megalinter/latest/descriptors/python_isort)                                        |  163|    0|                                                                          0|0.73s       |
|✅ PYTHON           |[mypy](https://oxsecurity.github.io/megalinter/latest/descriptors/python_mypy)                                          |  163|     |                                                                          0|119.01s     |
|✅ PYTHON           |[pylint](https://oxsecurity.github.io/megalinter/latest/descriptors/python_pylint)                                      |  163|     |                                                                          0|18.23s      |
|✅ REPOSITORY       |[git_diff](https://oxsecurity.github.io/megalinter/latest/descriptors/repository_git_diff)                              |yes  |     |no                                                                         |0.31s       |
|✅ REPOSITORY       |[secretlint](https://oxsecurity.github.io/megalinter/latest/descriptors/repository_secretlint)                          |yes  |     |no                                                                         |11.31s      |
|:warning: REPOSITORY|[semgrep](https://oxsecurity.github.io/megalinter/latest/descriptors/repository_semgrep)                                |yes  |     |[**182**](https://github.com/oxsecurity/megalinter/actions/runs/2728373587)|53.84s      |
|✅ REPOSITORY       |[trivy](https://oxsecurity.github.io/megalinter/latest/descriptors/repository_trivy)                                    |yes  |     |no                                                                         |27.87s      |
|✅ SPELL            |[cspell](https://oxsecurity.github.io/megalinter/latest/descriptors/spell_cspell)                                       |  868|     |                                                                          0|29.43s      |
|✅ SPELL            |[misspell](https://oxsecurity.github.io/megalinter/latest/descriptors/spell_misspell)                                   |  700|    0|                                                                          0|0.88s       |
|✅ XML              |[xmllint](https://oxsecurity.github.io/megalinter/latest/descriptors/xml_xmllint)                                       |    3|     |                                                                          0|0.01s       |
|✅ YAML             |[prettier](https://oxsecurity.github.io/megalinter/latest/descriptors/yaml_prettier)                                    |   79|    0|                                                                          0|3.45s       |
|✅ YAML             |[v8r](https://oxsecurity.github.io/megalinter/latest/descriptors/yaml_v8r)                                              |   22|     |                                                                          0|28.45s      |
|✅ YAML             |[yamllint](https://oxsecurity.github.io/megalinter/latest/descriptors/yaml_yamllint)                                    |   80|     |                                                                          0|1.33s       |


See errors details in [**artifact MegaLinter reports** on CI Job page](https://github.com/oxsecurity/megalinter/actions/runs/2728373587)

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 1]
```diff
@@ -29,19 +29,20 @@ linters:
     linter_url: https://github.com/dotnet/format
     linter_repo: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:33]: For dotnet 6.0 this shoudl be:
```yaml
    cli_lint_extra_args:
      - "--verify-no-changes"
      - "--exclude"
      - "/"
      - "--include"```

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 2]
```diff
@@ -29,19 +29,20 @@ linters:
...
       - "--folder"
       - "--check"
       - "--exclude"
       - "/"
       - "--include"
-    cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround for PythonBlackOrIsortLinter class behavior
+    cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround for DotnetFormatLinter class behavior
     cli_lint_fix_remove_args:
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:40]:
```yaml
    cli_lint_fix_remove_args:
      - "--verify-no-changes"
      ```

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 3]
```diff
@@ -28,6 +28,7 @@ linters:
     is_formatter: true
     linter_url: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:32]: Same for vb:
```yaml
    cli_lint_extra_args:
      - "--verify-no-changes"
      - "--exclude"
      - "/"
      - "--include"```

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 4]
```diff
@@ -39,8 +40,8 @@ linters:
       - "--check"
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:40]:
```yaml
    cli_lint_fix_remove_args:
      - "--verify-no-changes"
      ```

**Review by lextatic — CHANGES_REQUESTED** [review]
You probably need those changes and then it should work nicely.

**Comment by nvuillam** [comment]
@lextatic thanks for the tips , I made the updates :)

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 5]
```diff
@@ -29,19 +29,20 @@ linters:
...
-      - "--check"
+      - "--verify-no-changes"
     linter_version_cache: "000"
     examples:
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:43]: Sorry, I forgot the examples.

```yaml
examples:
      - "dotnet-format --verify-no-changes --exclude / --include myfile.cs myfile2.cs"
      - "dotnet-format --exclude / --include myfile.cs myfile2.cs" 
```

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 6]
```diff
@@ -28,19 +28,20 @@ linters:
...
+      - "--verify-no-changes"
       - "--exclude"
       - "/"
       - "--include"
     cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround to allow fix
     cli_lint_fix_remove_args:
-      - "--check"
+      - "--verify-no-changes"
     linter_version_cache: "000"
     examples:
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:42]:
```yaml
    examples:
      - "dotnet-format --verify-no-changes --exclude / --include myfile.vb myfile2.vb"
      - "dotnet-format --exclude / --include myfile.vb myfile2.vb"
```

**Review by lextatic — CHANGES_REQUESTED** [review]
Forgot to comment on the examples.

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 7]
```diff
@@ -29,19 +29,20 @@ linters:
     linter_url: https://github.com/dotnet/format
     linter_repo: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
       - "--folder"
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:34]: Remove `--folder`

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 8]
```diff
@@ -28,19 +28,20 @@ linters:
     is_formatter: true
     linter_url: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
       - "--folder"
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:33]: Remove `--folder`

**Review by lextatic — CHANGES_REQUESTED** [review]

**Comment by nvuillam** [comment]
Updates done :)

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 9]
```diff
@@ -29,19 +29,19 @@ linters:
     linter_url: https://github.com/dotnet/format
     linter_repo: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:33]:
```yaml
    cli_lint_fix_remove_args:
      - "whitespace"
      - "--folder"
      - "--verify-no-changes"
      - "--exclude"
      - "/"
      - "--include"
```

File: megalinter/descriptors/csharp.megalinter-descriptor.yml [hunk 10]
```diff
@@ -29,19 +29,19 @@ linters:
...
-    cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround for PythonBlackOrIsortLinter class behavior
+    cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround for DotnetFormatLinter class behavior
     cli_lint_fix_remove_args:
-      - "--check"
+      - "--verify-no-changes"
     linter_version_cache: "000"
     examples:
-      - "dotnet-format --folder --check --exclude / --include myfile.cs"
-      - "dotnet-format --folder --exclude / --include myfile.cs" # Fix
+      - "dotnet-format --verify-no-changes --exclude / --include myfile.cs myfile2.cs"
```
Comment (lextatic) [inline megalinter/descriptors/csharp.megalinter-descriptor.yml:43]:
```yaml
    examples:
      - "dotnet-format whitespace --folder --verify-no-changes --exclude / --include myfile.cs myfile2.cs"
      - "dotnet-format whitespace --folder --exclude / --include myfile.cs myfile2.cs" # Fix
```

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 11]
```diff
@@ -28,19 +28,19 @@ linters:
     is_formatter: true
     linter_url: https://github.com/dotnet/format
     linter_image_url: https://user-images.githubusercontent.com/9797472/61659851-6bbdc880-ac7d-11e9-95f7-d30c7de1a18a.png
+    cli_lint_mode: list_of_files
     cli_lint_extra_args:
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:32]:
```yaml
    cli_lint_extra_args:
      - "whitespace"
      - "--folder"
      - "--verify-no-changes"
      - "--exclude"
      - "/"
      - "--include"
```

File: megalinter/descriptors/vbdotnet.megalinter-descriptor.yml [hunk 12]
```diff
@@ -28,19 +28,19 @@ linters:
...
       - "--include"
     cli_lint_fix_arg_name: "--megalinter-fix-flag" # Workaround to allow fix
     cli_lint_fix_remove_args:
-      - "--check"
+      - "--verify-no-changes"
     linter_version_cache: "000"
     examples:
-      - "dotnet-format --folder --check --exclude / --include myfile.vb"
-      - "dotnet-format --folder --exclude / --include myfile.vb"
+      - "dotnet-format --verify-no-changes --exclude / --include myfile.vb myfile2.vb"
```
Comment (lextatic) [inline megalinter/descriptors/vbdotnet.megalinter-descriptor.yml:42]:
```yaml
    examples:
      - "dotnet-format whitespace --folder --verify-no-changes --exclude / --include myfile.vb myfile2.vb"
      - "dotnet-format whitespace --folder --exclude / --include myfile.vb myfile2.vb" # Fix
```

**Review by lextatic — COMMENTED** [review]
Sorry for making you go back and forth with the CLI arguments.

If you just want to upgrade dotnet to v6 I think those suggested changes should do it, but unfortunately they won't solve the [issue that originated this branch](https://github.com/oxsecurity/megalinter/issues/1642#issue-1315755238). That would require a more elaborated solution.

**Comment by nvuillam** [comment]
@lextatic why this "whitespace" ? Aren't there other types of formatting than just whitespaces ? ^^

**Comment by lextatic** [comment]
> @lextatic why this "whitespace" ? Aren't there other types of formatting than just whitespaces ? ^^

Yes, dotnet-format can formate styling too, but it won't work without a `.csproj` or `.sln` file. So if you want to lint only `.cs` files exactly as it used to in dotnet 5.0 you have to link whitespaces only with the `--folder` argument. `--folder` is option exclusive for linting with `whitespace`.

Here's the doc I found about it: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-format

**Comment by lextatic** [comment]
> @lextatic why this "whitespace" ? Aren't there other types of formatting than just whitespaces ? ^^

On a second thought, someone linting .NET should probably have a `.sln` or `.proj` file anyway, and you already have a `DEFAULT_WORKSPACE` var that could point megalinter in that direction.

So you could just add a `.sln` or `proj` file to your csharp and vb automated tests and go on with without changing anything, dropping the whitespace and --folder args altogether and having all of dotnet-format's functionality.

`dotnet format --verify-no-changes --exclude / --include myfile.cs myfile2.cs` would work nicely as it is now.

Thinking about it, that seems the most plausible solution, I was worried about making the new code pass your previous tests but in this case we should adapt the tests as the system changed. And this would be just add the `project` files.

**Comment by nvuillam** [comment]
@lextatic do you have an example of .sln file that would work ?

**Comment by lextatic** [comment]
> @lextatic do you have an example of .sln file that would work ?

Created a PR with example project files.

**Comment by lextatic** [comment]
It looks like DevSkim has not been upgraded to dotnet 6.0 yet.

https://github.com/microsoft/DevSkim
https://github.com/microsoft/DevSkim/pulls

**Comment by nvuillam** [comment]
:/
Do you know if it is possible to make dotnet 5 & dotnet 6 live together ? ( I never developed in dotnet)

**Comment by Kurt-von-Laven** [comment]
There may be other ways as well, but I know [the dotent-core asdf plugin](https://github.com/emersonsoares/asdf-dotnet-core) is capable of this.

**Comment by lextatic** [comment]
> :/ Do you know if it is possible to make dotnet 5 & dotnet 6 live together ? ( I never developed in dotnet)

Yes, I have both SDKs on my machine and they're interchangeable. You can have a `global.json` file on a folder and when dotnet is executed from this location it uses the SDK configured in the file.

Reference: https://www.markheath.net/post/switching-between-netcore-sdk-versions

**Comment by nvuillam** [comment]
Is there a command like `dotnet switch-to-v5` and `dotnet switch-to-v6` that I could easily use ?

**Comment by lextatic** [comment]
> Is there a command like `dotnet switch-to-v5` and `dotnet switch-to-v6` that I could easily use ?

Hmm, I guess `dotnet new globaljson --sdkversion <version> --force` does exactly that, you just have to check if both versions are installed under `.NET SDKs installed` upon using `dotnet --info`

For example:

`dotnet new globaljson --sdkversion 5.0.408 --force` to use dotnet 5.0
and
`dotnet new globaljson --sdkversion 6.0.302 --force` to use dotnet 6.0

**Comment by nvuillam** [comment]
@lextatic  Just asked DevSkim if they planned an upgrade, it would ease our task here ^^

https://github.com/microsoft/DevSkim/issues/391

**Comment by nvuillam** [comment]
@lextatic FYI :) https://github.com/microsoft/DevSkim/issues/391

**Pull request closed without merging**